### PR TITLE
Move the DOM memory management blog post into the tree.

### PR DESCRIPTION
**Pull request by Ms2ger**
This is a WIP, but I wanted to get eyes on it already.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/5745)

### CODE REVIEW

File: components/script/docs/JS-Servos-only-GC.md [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+```
+
+This macro describes which members of a C++ class should be added to a graph
+of potential cycles. Forgetting an entry can produce a memory leak. In Servo
+the consequences would be even worse: if the garbage collector can't see all
+references, it might free a node that is still in use. It's essential for both
+security and programmer convenience that we get rid of this manual listing of
+fields.
+
+Rust has a notion of [traits][traits], which are similar to
```
Comment (Manishearth) [inline components/script/docs/JS-Servos-only-GC.md]: nit: traits link broken

File: components/script/docs/JS-Servos-only-GC.md [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+document.rs:199     window.root_ref()
+document.rs:200 }
+document.rs:197:57: 200:6 <span style="color: green">note:</span> <b>...but borrowed value is only
+    valid for the block at 197:56</b>
+document.rs:197 fn bogus_get_window<"a>(&amp;self) -> JSRef<"a, Window> {
+document.rs:198     let window = self.window.root();
+document.rs:199     window.root_ref()
+document.rs:200 }</pre>
+
+We also implement the [`Deref` trait][deref] for `JSRef<T>`. This allows us to
```
Comment (Manishearth) [inline components/script/docs/JS-Servos-only-GC.md]: Might want to mention that due to Rust's autoderef accessing the inner fields is usually not a problem,

File: components/script/docs/JS-Servos-only-GC.md [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+[js-lint]: https://github.com/kmcallister/servo/commit/c20b50bbbbcdc8ce3551adbc1e039a727cf89995 (better link?)
+[warnings]: http://doc.rust-lang.org/rust.html#lint-check-attributes (better link?)
+
+Our plugin looks at every place where the code mentions a type. Remarkably,
+this adds only a fraction of a second to the compile time for Servo's largest
+subcomponent, as Rust compile times are dominated by [LLVM][llvm]'s back-end
+optimizations and code generation. The current version of the plugin is very
+simple and will miss some mistakes, like storing a struct containing `JS<T>` on
+the stack. (XXX does it still?) However, lint plugins run at a late stage of
```
Comment (Manishearth) [inline components/script/docs/JS-Servos-only-GC.md]: Such a struct can't be defined without being `must_root`, _unless_ it has generics, in which case there are ways of sidestepping the lint.

File: components/script/docs/JS-Servos-only-GC.md [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+// XXX: this gives me a crazy idea: what if we made JS<T> !Copy?
+
+This rule doesn't correspond to anything that already exists in Rust's type
+system. Fortunately, the Rust compiler can load 'lint plugins' providing custom
+static analysis. These basically take the form of new compiler warnings,
+although in this case we set the default severity to 'error'.
+
+[lints]: https://github.com/rust-lang/rust/pull/15024 (better link?)
+
+We have already [implemented a plugin][js-lint] which simply forbids `JS<T>`
```
Comment (Manishearth) [inline components/script/docs/JS-Servos-only-GC.md]: FWIW the current lint doesn't forbid `JS<T>`, I went a different route.

It forbids:

 - `must_root` types in non-`must_root` structs
 - assignment of local variables to `must_root` types
 - having functions take in `must_root` types

**Comment by jdm** [comment]
Reviewed files:
- components/script/docs/JS-Servos-only-GC.md @ r1

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 134 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5745#-JnRE5tCS0NHUSg0Nyvf)** ([raw file](https://github.com/servo/servo/blob/511ef22f5fd6147d57cf928672ab66eb16243aa2/components/script/docs/JS-Servos-only-GC.md#L134)):</sup>
Not sure what you have in mind here.

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 314 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5745#-JnREQ6PfIKgM_4dLC54)** ([raw file](https://github.com/servo/servo/blob/511ef22f5fd6147d57cf928672ab66eb16243aa2/components/script/docs/JS-Servos-only-GC.md#L314)):</sup>
You are correct. It does allow us to transition to a moving GC in the future, however.

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 342 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5745#-JnREkIIVjBxnSihCPhZ)** ([raw file](https://github.com/servo/servo/blob/511ef22f5fd6147d57cf928672ab66eb16243aa2/components/script/docs/JS-Servos-only-GC.md#L342)):</sup>
That might complicate our code that copies elements between rooted vectors.

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 396 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5745#-JnRFEShWeAmgUxDjuRu)** ([raw file](https://github.com/servo/servo/blob/511ef22f5fd6147d57cf928672ab66eb16243aa2/components/script/docs/JS-Servos-only-GC.md#L396)):</sup>
I don't think so; why is it important?

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by Manishearth** [comment]
Reviewed files:
- components/script/docs/JS-Servos-only-GC.md @ r2

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by jdm** [comment]
<sup>**[components/script/docs/JS-Servos-only-GC.md, line 173 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5745#-JoiIGdyvQ6YSv3BfnvE)** ([raw file](https://github.com/servo/servo/blob/beff2541a147588d54c4d9cbf0c6efeab77fff77/components/script/docs/JS-Servos-only-GC.md#L173)):</sup>
This will go stale.

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 239 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5745#-JoiIIz0XCeMigAnc_bA)** ([raw file](https://github.com/servo/servo/blob/beff2541a147588d54c4d9cbf0c6efeab77fff77/components/script/docs/JS-Servos-only-GC.md#L239)):</sup>
This will go stale.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by Manishearth** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed files:
- components/script/docs/JS-Servos-only-GC.md @ r3

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by metajack** [comment]
@ms2ger Still working on this?

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 
Still some XXX bits remaining.

---

Reviewed 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 257 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5745#-JvrGhdip3ucHimWIxlW)** ([raw file](https://github.com/servo/servo/blob/086049c8cddb34138d2c8bfa8b9889d3f8ab1037/components/script/docs/JS-Servos-only-GC.md#L257)):</sup>
?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by Manishearth** [comment]
(this also needs updating post-jsref)

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 


---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 230 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5745#-K19pDKrWAn7tahdFn5p)** ([raw file](https://github.com/servo/servo/blob/72f3a8e9b7f63071cf8a54dde004a6de3c44df21/components/script/docs/JS-Servos-only-GC.md#L230)):</sup>
We can now dereference them without rooting.


---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 316 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5745#-K19u2hA-2vU20NEURku)** ([raw file](https://github.com/servo/servo/blob/72f3a8e9b7f63071cf8a54dde004a6de3c44df21/components/script/docs/JS-Servos-only-GC.md#L316)):</sup>
Perhaps 'effectively' instead of 'simply'?


---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 348 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5745#-K19tl1wNitGhTw8xPfM)** ([raw file](https://github.com/servo/servo/blob/72f3a8e9b7f63071cf8a54dde004a6de3c44df21/components/script/docs/JS-Servos-only-GC.md#L348)):</sup>
s/collection/collected/


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by Manishearth** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 30 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5745#-K1AWpfk4sWpPKBmHYND)** ([raw file](https://github.com/servo/servo/blob/72f3a8e9b7f63071cf8a54dde004a6de3c44df21/components/script/docs/JS-Servos-only-GC.md#L30)):</sup>
low level object where? Rust-side/Spidermonkey-side isn't very clear here


---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 69 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5745#-K1AXjuOg7wtYdwHn5C3)** ([raw file](https://github.com/servo/servo/blob/72f3a8e9b7f63071cf8a54dde004a6de3c44df21/components/script/docs/JS-Servos-only-GC.md#L69)):</sup>
Isn't "manually breaking possible cycles" the same as cycle collection?


---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 319 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5745#-K1AdDHACguUObCgFte8)** ([raw file](https://github.com/servo/servo/blob/72f3a8e9b7f63071cf8a54dde004a6de3c44df21/components/script/docs/JS-Servos-only-GC.md#L319)):</sup>
We don't allow it in DOM struct definitions. The lint explicitly only emits the error on structs without a must_root annotation. (DOM structs have this annotation)


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by Manishearth** [comment]
Reviewed 1 of 1 files at r1, 1 of 2 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by Manishearth** [comment]
Reviewed 1 of 2 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by Ms2ger** [comment]
Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 351 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5745#-JnI5B2ccJpPuj0fwf6V-r1-351)** ([raw file](https://github.com/servo/servo/blob/511ef22f5fd6147d57cf928672ab66eb16243aa2/components/script/docs/JS-Servos-only-GC.md#L351)):</sup>
Do you want to write it out?


---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 69 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5745#-K1AXjuOg7wtYdwHn5C3)** ([raw file](https://github.com/servo/servo/blob/72f3a8e9b7f63071cf8a54dde004a6de3c44df21/components/script/docs/JS-Servos-only-GC.md#L69)):</sup>
No, "manually breaking possible cycles" means custom code for each concrete field.


---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 230 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5745#-K19pDKrWAn7tahdFn5p)** ([raw file](https://github.com/servo/servo/blob/72f3a8e9b7f63071cf8a54dde004a6de3c44df21/components/script/docs/JS-Servos-only-GC.md#L230)):</sup>
Should I just drop the note about `JS::root`, then? Maybe we should remove that method (in favour of `(&T).root()`)?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by Manishearth** [comment]
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 351 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5745#-JnI5B2ccJpPuj0fwf6V-r1-351)** ([raw file](https://github.com/servo/servo/blob/511ef22f5fd6147d57cf928672ab66eb16243aa2/components/script/docs/JS-Servos-only-GC.md#L351)):</sup>
Just mention "forbids JS<T> from appearing on the stack"


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 


---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/docs/JS-Servos-only-GC.md, line 230 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5745#-K19pDKrWAn7tahdFn5p)** ([raw file](https://github.com/servo/servo/blob/72f3a8e9b7f63071cf8a54dde004a6de3c44df21/components/script/docs/JS-Servos-only-GC.md#L230)):</sup>
Both of those things sound sensible to me.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Comment by Manishearth** [comment]
@bors-servo r+


---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5745)

**Pull request merged**